src/Chapter3.hs

```diff
@@ -480,6 +514,34 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data OptionalCastle 
+  = NoCastle
+  | Castle 
+    { name :: String
+    , wall :: Bool} deriving (Show)
```

Comment:
The idea is absolutely correct in here 👌🏼 
However, having records in sum data types are considered a bad practice and better to be avoided, as it created partial functions, unfortunately. 
In normal record `name` function would be `name :: OptionalCastle -> String`. Here it would have the same type, but if you use it on `NoCastle` it will fail.

There are a number of options here to fix this.
1. You can instead of using `Bool` as the state indfecator of the wall, add separate constructors for both options:
    ```haskell
    data OptionalCastle
        = NoCastle
        | Castle String
        | CastleWithWalls String
    ```
2. You can move the record into a separate type and use it in the sum type:
```haskell
data OptionalCastle 
  = NoCastle
  | Castle CastleStructure
    
data CastleStructure = CastleStructure
    { name :: String
    , wall :: Bool
    } deriving (Show)
```